consensus/hotstuff/safetyrules/safety_rules_test.go

```diff
@@ -1,165 +1,526 @@
+
+		timeout, err := s.safety.ProduceTimeout(view, newestQC, nil)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), timeout)
+	})
+	s.Run("cur-view-below-highest-acknowledged-view", func() {
+		view := s.safetyData.HighestAcknowledgedView
+		newestQC := helper.MakeQC(helper.WithQCView(s.safetyData.LockedOneChainView))
+
+		timeout, err := s.safety.ProduceTimeout(view-2, newestQC, nil)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), timeout)
+	})
+	s.Run("cur-view-below-highest-QC", func() {
+		newestQC := helper.MakeQC(helper.WithQCView(s.safetyData.LockedOneChainView))
+		view := newestQC.View
+
+		timeout, err := s.safety.ProduceTimeout(view, newestQC, nil)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), timeout)
+	})
+
+	s.signer.AssertNotCalled(s.T(), "CreateTimeout")
+	s.signer.AssertNotCalled(s.T(), "PutSafetyData")
+}
 
-	// create voter
-	block, _, voter := createVoter(blockView, lastVotedView, isBlockSafe, isCommitteeMember)
+// TestProduceTimeout_CreateTimeoutException tests that no timeout is created if timeout creation raised an exception
+func (s *SafetyRulesTestSuite) TestProduceTimeout_CreateTimeoutException() {
+	view := s.proposal.Block.View
+	newestQC := helper.MakeQC(helper.WithQCView(view - 1))
+
+	exception := errors.New("create-vote-exception")
+	s.signer.On("CreateTimeout", view, newestQC, (*flow.TimeoutCertificate)(nil)).Return(nil, exception).Once()
+	vote, err := s.safety.ProduceTimeout(view, newestQC, nil)
+	require.Nil(s.T(), vote)
+	require.ErrorIs(s.T(), err, exception)
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
+}
 
-	// produce vote
-	_, err := voter.ProduceVote(block, curView)
-	require.NoError(t, err)
+// TestProduceTimeout_PersistStateException tests that no timeout is created if persisting state failed
+func (s *SafetyRulesTestSuite) TestProduceTimeout_PersistStateException() {
+	exception := errors.New("persister-exception")
+	s.persister.On("PutSafetyData", mock.Anything).Return(exception)
+
+	view := s.proposal.Block.View
+	newestQC := helper.MakeQC(helper.WithQCView(view - 1))
+	expectedTimeout := &model.TimeoutObject{
+		View:     view,
+		NewestQC: newestQC,
+	}
 
-	// produce vote again for the same view
-	_, err = voter.ProduceVote(block, curView)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "must be larger than the last voted view")
-	require.False(t, model.IsNoVoteError(err))
+	s.signer.On("CreateTimeout", view, newestQC, (*flow.TimeoutCertificate)(nil)).Return(expectedTimeout, nil).Once()
+	timeout, err := s.safety.ProduceTimeout(view, newestQC, nil)
+	require.Nil(s.T(), timeout)
+	require.ErrorIs(s.T(), err, exception)
 }
 
-func testVotingWhileNonCommitteeMember(t *testing.T) {
-	blockView, curView, lastVotedView, isBlockSafe, isCommitteeMember := uint64(3), uint64(3), uint64(2), true, false
+// TestProduceTimeout_AfterVote tests a case where we first produce a timeout and then try to vote
```

Comment:
This doc seems to be the opposite of the implementation (maybe got mixed with the doc for the other suggested test case)

Reply:
Thanks, I've messed up when was reordering and writing godoc.